SQL/advanced/triggers/02_before_after.sql

```diff
@@ -0,0 +1,101 @@
+/*******************************************
+Difference between BEFORE and AFTER triggers
+*******************************************/
+
+DROP FUNCTION IF EXISTS
+  tr_delete_client, tr_search_name
+CASCADE;
+
+/********************
+BEFORE is used when we need to validate data before any action on it
+
+For example, we want to check if we can delete some rows from the table.
+AFTER trigger cannot be used because we've already deleted the rows for us.
+********************/
+CREATE FUNCTION tr_delete_client()
+  RETURNS trigger
+  LANGUAGE plpgsql
+AS $$
+/**
+Check client's status before deletion
+*/
+DECLARE
+  _clients_status varchar;
+
+BEGIN
+  SELECT
+    status
+  INTO STRICT
+    _clients_status
+  FROM
+    clients
+  WHERE
+    id = OLD.id;
+
+  IF _clients_status = 'VIP' THEN
+    RAISE EXCEPTION 'The precious client % cannot leave our company!', OLD.id;
+  END IF;
+
+  RETURN OLD;
+END;
+$$;
+
+CREATE TRIGGER client_tr_delete
+  BEFORE DELETE ON clients
```

Comment:
what are other use cases for BEFORE and crucial difference from AFTER?

Reply:
In a `BEFORE` trigger, the trigger is executed BEFORE the DML statement is executed. So we have a chance to modify the `NEW` row (`OLD` in case of DELETE) row BEFORE it is inserted/updated/deleted.

In an AFTER trigger, the trigger is executed AFTER the DML statement is executed. We have already lost the ability to modify the `NEW` and `OLD` records and have it actually mean anything. However, we can still do whatever we want in the trigger like raising exceptions, doing other inserts/updates/deletes, etc.

Reply:
The point regarding modification before insert / update is correct and this is one of the most crucial differences. I don't really know real-life example of modification OLD before deletion...

Os I would say we have about 3 main point that affect (or not) the choice:
1. The need for modification (BEFORE INSERT / UPDATE only)
2. The need to raise an exception in some circumstances (ANY of the triggers)
3. Work with related objects (AFTER INSERT, BEFORE DELETE and maybe BEFORE/AFTER UPDATE if we need to work with a change of relation).

Reply:
Got it!

Then, I'd like to create different triggers for each case after we resolve two points above.

Reply:
1. Done
2. Done, in my cases `BEFORE` is just more efficient.
3. Done.